api/localizer/writer.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+// See proposals/localize-command
+package localizer
+
+import (
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/kustomize/api/loader"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+)
+
+const urlSeparator = "/"
+const localizeDir = "localized-files"
+const alpha = "alpha"
+
+var errNewDirRef = errors.New("target references newDir")
+var errAbsPath = errors.New("absolute paths not yet supported")
+
+func makeAndClean(path string, fSys filesys.FileSystem) (filesys.ConfirmedDir, error) {
+	var cleanPath filesys.ConfirmedDir
+	var err error
+
+	if err = fSys.MkdirAll(path); err != nil {
+		return "", errors.Wrapf(err, filesys.ErrMkDirAll.Error())
+	}
+	if cleanPath, _, err = fSys.CleanedAbs(path); err != nil {
+		return "", errors.Wrapf(err, filesys.ErrCleanAbs.Error())
+	}
+
+	return cleanPath, nil
+}
+
+// NewWriter is the factory method for Writer.
+func NewWriter(target filesys.ConfirmedDir, scope filesys.ConfirmedDir, newDir filesys.ConfirmedDir, fSys filesys.FileSystem) (*Writer, error) {
+	inNewDir, _ := filepath.Rel(scope.String(), target.String())
+	dst, dstErr := makeAndClean(newDir.Join(inNewDir), fSys)
+	if dstErr != nil {
+		return nil, errors.Wrapf(dstErr, "unable to write to newDir")
+	}
+
+	return &Writer{
+		fSys:   fSys,
+		newDir: newDir,
+		root:   target,
+		dstDir: dst,
+	}, nil
+}
+
+// Writer takes care of writing localized files to their appropriate destination
+// in newDir for command kustomize localize. Each Writer is associated with a
+// kustomization root, either remote or local, referenced by target and writes
+// files that the root directly references.
+//
+// Writer is meant to work with loader, which performs the path checks.
+type Writer struct {
```

Comment:
nit: Put this struct definition before any uses of it (so at least above NewWriter)

Also, do this and NewWriter need to be public?

Reply:
> Also, do this and NewWriter need to be public?

I made them public to follow the `kusttarget` and `Loader` example.  

https://github.com/kubernetes-sigs/kustomize/blob/master/api/internal/target/kusttarget.go
https://github.com/kubernetes-sigs/kustomize/blob/master/api/loader/loader.go#L19

The file-specific implementation details are abstracted out of `kusttarget` into `Loader`, which is a public interface under `api` > `loader`. 

I planned to create a `loctarget` that calls `NewWriter` and the `Writer` methods. However, I also understand that the localize `Writer` is not as reusable as `Loader`. If I don't need to follow the `Loader` structure, I'd be happy to put all localize files inside a single directory.

Reply:
We don't need to follow the `Loader` structure and in general these structs/methods seem to be very specific to `localize`, so I think it would be best to put them all in the same package and only expose what's necessary. So yeah, when you create a `loctarget`, let's put it in the same directory.

Reply:
I plan to put them all in the `localizer` package under `api/internal`. I could've left everything private in `api/localizer`, but `lint` requires the test files be in a different package, so I had to expose the types and put them all under `internal`.